File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.js
function solution(n, costs) {
  costs.sort((a, b) => a[2] - b[2]);
  const arr = Array.from({ length: n }, (_, i) => ({ cost: 0, root: i }));
  let cost = 0;

  for (let i = 0; i < costs.length; i += 1) {
    if (union(arr, costs[i][0], costs[i][1])) {
      cost += costs[i][2];
    }
  }

  return cost
}

function union(arr, x, y) {
  const xr = find(arr, x);
  const yr = find(arr, y);

  arr[yr].root = arr[xr].root;
  return xr !== yr;
}

function find(arr, x) {
  let cur = x;
  while (cur !== arr[cur].root) {
    cur = arr[cur].root;
  }
  return cur;
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;

class Solution {
    int[] graph;
    
    public int solution(int n, int[][] costs) {
        graph = new int[n];
        
        for(int i=0; i<n; i++) {
            graph[i] = i;
        }
        
        Arrays.sort(costs, (o1, o2)->Integer.compare(o1[2], o2[2]));
        
        int answer = 0;
        for(int[] cost: costs) {
            int x = cost[0];
            int y = cost[1];
            int c = cost[2];
            
            if (union(x, y)) {
                answer += c;
            }
        }
        
        return answer;
    }
    
    private int find(int x) {
        if (graph[x] != x) {
            graph[x] = find(graph[x]);
        }
        return graph[x];
    }
    
    private boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        
        if (rootX == rootY) return false;
        
        graph[rootY] = rootX;
        return true;
    }
}